test/integration/fake_upstream.cc

```diff
@@ -147,7 +147,18 @@ void FakeConnectionBase::close() {
 }
 
 void FakeConnectionBase::readDisable(bool disable) {
-  connection_.dispatcher().post([this, disable]() -> void { connection_.readDisable(disable); });
+  std::unique_lock<std::mutex> lock(lock_);
+  if (!disconnected_) {
+    connection_.dispatcher().post([this, disable]() -> void {
+      if (!disconnected_) {
+        connection_.readDisable(disable);
+      }
```

Comment:
I think the changes in this function are still trying to fix a previous race condition. Should we remove them fro this PR or just RELEASE_ASSERT(!disconnected_) after taking the lock?